Replace TxBody with a MemoBytes and backward compatible Coders.hs encoding

TimSheard opened the pull request:
Made extensive changes to Coders.hs to handle sparse encoding. Applied these
changes to TxBody, so that it is now defined as a sparsely encoded MemoBytes.

It is probably best not to look to closely at Data.Coders.hs as it has undergone extensive changes.

TimSheard committed:
"Added the TxBody type with validity intervals and forge fields. Tied this
together with the Timelocks scripts. TxBody is newtype wrapped around a MemoBytes.
It exports a set of HasField instances appropriate for a TxBody. Also created a
test file with a minumum number of test, testing the HasField use and roundtripp
CBOR properties. Cleaned up and extended Jareds idea of using EraIndependentTxBody.
Extended this to all types with a HashAnnotated instance, which now adds an
associated type family HashIndex. Made all uses consistent with this approach,
Removed the function eraIndTxBodyHash, which can be replaced hashAnnotated."

TimSheard committed:
Extended Coder.hs to handle redefining TxBody as sparse encoded MemoBytes.

Made extensive changes to Coders.hs to handle sparse encoding. Applied these
changes to TxBody, so that it is now defined as a sparsely encoded MemoBytes

JaredCorduan commented on shelley-ma/shelley-ma-test/src/Test/Cardano/Ledger/ShelleyMA/TxBody.hs:
```diff
@@ -103,12 +103,12 @@ fieldTests =
     [ testCase "inputs" (assertEqual "inputs" (getField @"inputs" tx) empty),
       testCase "outputs" (assertEqual "outputs" (getField @"outputs" tx) eseq),
-      testCase "dcerts" (assertEqual "dcerts" (getField @"certs" tx) eseq),
+      testCase "certs" (assertEqual "certs" (getField @"certs" tx) eseq),
```
~did you mean rename this?~ nm, makes sense. I guess we have some inconsistencies between `dcert` and `cert` already...

TimSheard replied:
try to make consistent with Nicks uses.

JaredCorduan commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/TxBody.hs:
```diff
@@ -39,17 +42,17 @@ module Shelley.Spec.Ledger.TxBody
         _ttl,
         _txUpdate,
-        _mdHash,
-        extraSize
```
thank you for removing this!

TimSheard replied:
my pleasure

redxaxder commented on semantics/executable-spec/src/Data/Coders.hs:
Maybe add an explanation here. What is thin?

TimSheard replied:
Good call, much cleaner now.

redxaxder commented on semantics/executable-spec/src/Data/Coders.hs:
This could use some more explanation

TimSheard replied:
add comments about how the 3 different kinds of constructors are used.

redxaxder commented on semantics/executable-spec/src/Data/Coders.hs:
It seems like this one could be a function instead of a GADT constructor

TimSheard replied:
I agree,it could be a function
ed (Dual enc dec) x = E enc x
But that obscures the nice visual inspection of Duality, between (ED dual t) and (DD dual)
so lets leave it for now.

redxaxder commented on semantics/executable-spec/src/Data/Coders.hs:
These constraints might not need to be here

TimSheard replied:
Yes, they can be removed.

redxaxder commented on semantics/executable-spec/src/Data/Coders.hs:
In the future some of our keys might not be Words

TimSheard replied:
We can keep that in mind. If that was so, we'd need some kind of class here, witnessing what types can be keys.
We can add that if and when we need it.

redxaxder commented on semantics/executable-spec/src/Data/Coders.hs:
If the `Wrapped` type was higher kinded, it might be possible to eliminate this

redxaxder commented on semantics/executable-spec/src/Data/Coders.hs:
This could also be a function instead of a GADT constructor

redxaxder commented on semantics/executable-spec/src/Data/Coders.hs:
```diff
@@ -320,56 +421,160 @@ decodE x = decodeCount x 0
+  y <- decodeClosed x
+  pure (i,f <*> y)
+
+decodeClosed :: Decode w t -> Decoder s t
```
Given this type, It doesn't seem impossible for `SumD` to appear here. Did you mean to have a constraint?

TimSheard replied:
By adding a Proxy to the args of decodeClosed, I can make the function have an explicit (Applys w1 w2) constraint.         Applys w1 w2 => Proxy w1 -> Decode w2 t -> Decoder s t
Since there is no (Applys x 'Thin) instance the SumC case is now truly unreachable. I also made sure that decodeClosed is not exported. So no one knows it exists. But even if they did know, they could not call it except in a context where the (Applys w1 w2) constraint is in scope.

redxaxder commented on semantics/executable-spec/src/Data/Coders.hs:
```diff
@@ -320,56 +421,160 @@ decodE x = decodeCount x 0
-encodeList :: ToCBOR a => [a] -> Encoding
-encodeList = encodeFoldable
+dualMaybe :: (ToCBOR a, FromCBOR a) => Dual (Maybe a)
+dualMaybe = Dual toCBOR fromCBOR
```
The encoding of `Maybe` provided by the `ToCBOR` instance encodes it as an empty or nonempty list. `[a] / []`.

The encoding we've been using is `a / null`.

TimSheard replied:
We made two different Duals for Maybe
dualMaybeAsList
dualMaybeAsNull

redxaxder commented on semantics/executable-spec/src/Data/MemoBytes.hs:
```diff
@@ -87,6 +94,11 @@ printMemo x = putStrLn (showMemo x)
+memoBytesAnn t = Memo term (shorten bytes)
+   where bytes = toLazyByteString (encode t)
+         Annotator f = runE t
+         term = f (Full bytes)
```
The `Annotator` type is produced when decoding. I would not expect it to ever appear in an encoder like this.

Since this appears to not be used anywhere at the moment, and locations where it is used are likely to have arrived at the need for it by mistake, I suggest removing it.

TimSheard replied:
Yes this should have been removed earlier.

redxaxder commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/TxBody.hs:
```diff
@@ -21,6 +22,8 @@
 {-# LANGUAGE UndecidableInstances #-}
 {-# LANGUAGE ViewPatterns #-}
+{-# OPTIONS_GHC -fno-warn-unused-binds #-}
+{-# OPTIONS_GHC -fno-warn-unused-imports #-}
```
The warnings are off.

TimSheard replied:
Deleted turning off the warnings.

redxaxder commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/TxBody.hs:
```diff
@@ -1076,6 +1117,8 @@ instance
           _poolMD = maybeToStrictMaybe md
         }
 
+{-
```
These are commented out. Are they unused? Can they be deleted?

TimSheard replied:
Those are remnants of the switch to set algebras, Haven't been used in a long time.

TimSheard commented:
Addressed all of the comments. The idea of using higher order indexes, cleaned up the code considerably. See Density as an index to the index Wrapped.

redxaxder commented:
lgtm!
